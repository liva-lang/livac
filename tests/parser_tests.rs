use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper para casos correctos del parser
fn test_parser_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();
    
    // Convertir AST a JSON para snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("ok_{}.ast", test_name), json);
}

/// Test helper para casos de error del parser
fn test_parser_err(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/err_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: err_{}.liva", test_name));
    
    let tokens = tokenize(&source).unwrap();
    let result = parse(tokens, &source);
    assert!(result.is_err(), "Expected parser error for test: {}", test_name);
    
    let error_msg = result.unwrap_err().to_string();
    assert_snapshot!(format!("err_{}.diag", test_name), error_msg);
}

#[test]
fn test_functions_oneliner() {
    test_parser_ok("functions_oneliner");
}

#[test]
fn test_functions_block() {
    test_parser_ok("functions_block");
}

#[test]
fn test_classes() {
    test_parser_ok("classes");
}

#[test]
fn test_control_flow() {
    test_parser_ok("control_flow");
}

#[test]
fn test_expressions() {
    test_parser_ok("expressions");
}

#[test]
fn test_imports() {
    test_parser_ok("imports");
}

#[test]
fn test_unclosed_paren() {
    test_parser_err("unclosed_paren");
}

#[test]
fn test_unclosed_brace() {
    test_parser_err("unclosed_brace");
}

#[test]
fn test_case_without_switch() {
    test_parser_err("case_without_switch");
}

#[test]
fn test_return_outside_function() {
    test_parser_err("return_outside_function");
}

#[test]
fn test_duplicate_default() {
    test_parser_err("duplicate_default");
}
