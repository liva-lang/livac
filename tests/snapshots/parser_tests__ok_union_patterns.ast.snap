---
source: tests/parser_tests.rs
expression: json
---
{
  "items": [
    {
      "Function": {
        "name": "main",
        "type_params": [],
        "params": [],
        "return_type": null,
        "body": {
          "stmts": [
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "x"
                    },
                    "type_ref": {
                      "Union": [
                        {
                          "Simple": "int"
                        },
                        {
                          "Simple": "string"
                        }
                      ]
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 42
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "result1"
                    },
                    "type_ref": null
                  }
                ],
                "init": {
                  "Switch": {
                    "discriminant": {
                      "Identifier": "x"
                    },
                    "arms": [
                      {
                        "pattern": {
                          "Typed": {
                            "name": "n",
                            "type_ref": {
                              "Simple": "int"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Binary": {
                              "op": "Mul",
                              "left": {
                                "Identifier": "n"
                              },
                              "right": {
                                "Literal": {
                                  "Int": 2
                                }
                              }
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "s",
                            "type_ref": {
                              "Simple": "string"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "Int": 0
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "y"
                    },
                    "type_ref": {
                      "Union": [
                        {
                          "Simple": "int"
                        },
                        {
                          "Simple": "string"
                        },
                        {
                          "Simple": "bool"
                        }
                      ]
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "String": "hello"
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "result2"
                    },
                    "type_ref": null
                  }
                ],
                "init": {
                  "Switch": {
                    "discriminant": {
                      "Identifier": "y"
                    },
                    "arms": [
                      {
                        "pattern": {
                          "Typed": {
                            "name": "n",
                            "type_ref": {
                              "Simple": "int"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "String": "Number"
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "s",
                            "type_ref": {
                              "Simple": "string"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "String": "String"
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "b",
                            "type_ref": {
                              "Simple": "bool"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "String": "Boolean"
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "value"
                    },
                    "type_ref": {
                      "Union": [
                        {
                          "Simple": "int"
                        },
                        {
                          "Simple": "string"
                        }
                      ]
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 100
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "doubled"
                    },
                    "type_ref": null
                  }
                ],
                "init": {
                  "Switch": {
                    "discriminant": {
                      "Identifier": "value"
                    },
                    "arms": [
                      {
                        "pattern": {
                          "Typed": {
                            "name": "n",
                            "type_ref": {
                              "Simple": "int"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Binary": {
                              "op": "Add",
                              "left": {
                                "Identifier": "n"
                              },
                              "right": {
                                "Identifier": "n"
                              }
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "s",
                            "type_ref": {
                              "Simple": "string"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "Int": 0
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "multi"
                    },
                    "type_ref": {
                      "Union": [
                        {
                          "Simple": "int"
                        },
                        {
                          "Simple": "string"
                        },
                        {
                          "Simple": "bool"
                        }
                      ]
                    }
                  }
                ],
                "init": {
                  "Literal": {
                    "Int": 42
                  }
                },
                "is_fallible": false
              }
            },
            {
              "VarDecl": {
                "bindings": [
                  {
                    "pattern": {
                      "Identifier": "nested"
                    },
                    "type_ref": null
                  }
                ],
                "init": {
                  "Switch": {
                    "discriminant": {
                      "Identifier": "multi"
                    },
                    "arms": [
                      {
                        "pattern": {
                          "Typed": {
                            "name": "n",
                            "type_ref": {
                              "Simple": "int"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Switch": {
                              "discriminant": {
                                "Binary": {
                                  "op": "Gt",
                                  "left": {
                                    "Identifier": "n"
                                  },
                                  "right": {
                                    "Literal": {
                                      "Int": 50
                                    }
                                  }
                                }
                              },
                              "arms": [
                                {
                                  "pattern": {
                                    "Literal": {
                                      "Bool": true
                                    }
                                  },
                                  "guard": null,
                                  "body": {
                                    "Expr": {
                                      "Literal": {
                                        "String": "Large"
                                      }
                                    }
                                  }
                                },
                                {
                                  "pattern": {
                                    "Literal": {
                                      "Bool": false
                                    }
                                  },
                                  "guard": null,
                                  "body": {
                                    "Expr": {
                                      "Literal": {
                                        "String": "Small"
                                      }
                                    }
                                  }
                                }
                              ]
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "s",
                            "type_ref": {
                              "Simple": "string"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "String": "Text"
                            }
                          }
                        }
                      },
                      {
                        "pattern": {
                          "Typed": {
                            "name": "b",
                            "type_ref": {
                              "Simple": "bool"
                            }
                          }
                        },
                        "guard": null,
                        "body": {
                          "Expr": {
                            "Literal": {
                              "String": "Bool"
                            }
                          }
                        }
                      }
                    ]
                  }
                },
                "is_fallible": false
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    }
  ]
}
